ddl/ddl_api.go

```diff
@@ -5177,62 +5233,69 @@ func checkPlacementSpecs(specs []*ast.PlacementSpec) ([]*placement.Rule, error)
-		case ast.PlacementRoleLearner:
-			rule.Role = placement.Learner
-		case ast.PlacementRoleVoter:
-			rule.Role = placement.Voter
-		default:
-			return rules, errors.Errorf("invalid placement spec[%d], unknown role: %d", k, spec.Role)
-		}
-
-		for _, label := range strings.Split(spec.Constraints, ",") {
-			label = strings.TrimSpace(label)
-
-			if len(label) < 4 {
-				return rules, errors.Errorf("invalid placement spec[%d], constraint too short to be valid: %s", k, label)
-			}
-
-			var op placement.LabelConstraintOp
-			switch label[0] {
-			case '+':
-				op = placement.In
-			case '-':
-				op = placement.NotIn
+			err = errors.Errorf("unknown action type: %d", spec.Tp)
+		}
+
+		if err == nil {
+			switch spec.Role {
+			case ast.PlacementRoleFollower:
+				rule.Role = placement.Follower
+			case ast.PlacementRoleLeader:
+				rule.Role = placement.Leader
+			case ast.PlacementRoleLearner:
+				rule.Role = placement.Learner
+			case ast.PlacementRoleVoter:
+				rule.Role = placement.Voter
 			default:
-				return rules, errors.Errorf("invalid placement spec[%d], unknown operation: %c", k, label[0])
+				err = errors.Errorf("unknown role: %d", spec.Role)
 			}
+		}
 
-			kv := strings.Split(label[1:], "=")
-			if len(kv) != 2 {
-				return rules, errors.Errorf("invalid placement spec[%d], invalid constraint format: %s", k, label)
-			}
+		if err == nil {
+			cnstr := strings.TrimSpace(spec.Constraints)
 
-			key := strings.TrimSpace(kv[0])
-			if key == "" {
-				return rules, errors.Errorf("invalid placement spec[%d], empty constraint key: %s", k, label)
+			if len(cnstr) > 0 && cnstr[0] == '[' {
+				constraints := []string{}
+				err = json.Unmarshal([]byte(cnstr), &constraints)
+				if err == nil {
+					err = checkPlacementLabelConstraints(rule, constraints)
+					if err == nil {
+						rules = append(rules, rule)
+					}
+				}
+			} else if len(cnstr) > 0 && cnstr[0] == '{' {
+				constraints := map[string]int{}
+				err = json.Unmarshal([]byte(cnstr), &constraints)
+				if err == nil {
+					for labels, cnt := range constraints {
+						newrule := &placement.Rule{}
+						*newrule = *rule
+						if cnt <= 0 {
+							err = errors.Errorf("count should be non-positive, but got %d", cnt)
+							break
+						}
+						// TODO: handle or remove it in later commits
+						rule.Count -= cnt
+						newrule.Count = cnt
+						err = checkPlacementLabelConstraints(newrule, strings.Split(strings.TrimSpace(labels), ","))
+						if err != nil {
+							break
+						}
+						rules = append(rules, newrule)
+					}
+				}
+			} else {
+				err = errors.Errorf("constraint should be a JSON array or object, but got '%s'", cnstr)
 			}
+		}
 
-			val := strings.TrimSpace(kv[1])
-			if val == "" {
-				return rules, errors.Errorf("invalid placement spec[%d], empty constraint val: %s", k, label)
+		if err != nil {
+			sb.Reset()
+			if e := spec.Restore(restoreCtx); e != nil {
+				return rules, ErrInvalidPlacementSpec.GenWithStackByArgs(e, err)
```

Comment:
It doesn't make sense to replace the statement with `e`.
```suggestion
				return rules, ErrInvalidPlacementSpec.GenWithStackByArgs("", err)
```